interpreter/attributes.go

```diff
@@ -0,0 +1,188 @@
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// Attribute refers to a Variable and zero or more field or index traversals.
+type Attribute interface {
+	// Variable referenced by the attribute.
+	//
+	// When the Attribute is absolute, the Variable value is non-nil. When the Attribute is
+	// relative such as might happen when refering to the field or index of a object-based
+	// return value from a function, the Variable value will be nil.
+	Variable() Variable
+
+	// Path of field traversals against reference object, either a variable or a contextual
+	// intermediate value computed during CEL evaluation.
+	//
+	// If Variable is nil, the Path must contain at least one PathElem.
+	Path() []*PathElem
+
+	// Select a path element from the current Attribute to produce a more fully qualified
+	// Attribute.
+	Select(*PathElem) Attribute
+}
+
+// Variable refers to a type-level expression identifier corresponding to program input.
+type Variable interface {
+	// ID indicates the expression id where the variable appears.
+	//
+	// Non-zero when the Variable is identified during program plan time.
+	//
+	// Zero-ish when the Variable is provided as part of a user-provided Attribute.
+	ID() int64
+
+	// Name of the variable, may be simple or namespaced, e.g. 'data' or 'cloud.iam.resource'.
+	Name() string
+}
+
+// PathElem representing a fragment of the attribute selection path.
+type PathElem struct {
+	// ID of the expression where the PathElem occurs, non-zero if determined during program plan
+	// time.
+	ID int64
+
+	// ToValue function which resolve the PathElem to a concrete ref.Val.
+	//
+	// In most the PathElem will be a constant value; however, there are cases when an Attribute
+	// may be cross-referenced, e.g. a[b] where the Attribute of {var: a, path: [b]} depends on
+	// the resolution of 'b' before the Attribute reference is fully known. In these cases, the
+	// ToValue function reference permits flexible resolution of the cross-reference.
+	ToValue func(Activation) ref.Val
+}
+
+// NewUnknownAttribute creates an Attribute that refers to a top-level variable or one of its
+// descendants fields in a selection path.
+//
+// A selection path element may be of the following scalar types:
+//
+// - bool, types.Bool
+// - int, int32, int64, types.Int
+// - string, types.String
+// - uint, uint32, uint64, types.Uint
+//
+// The string path element '*' is treated as a wildcard match against any path element value.
+//
+// As absolute Attribute references within an expression are being resolved, the attribute path
+// is compared to the set of declared unknowns which refer to the same top-level variable. When
+// the unknown Attribute matches any part of the resolved attribute, a types.Unknown value is
+// generated. For example, when a top-level variable contains an unknown attribute, then it too
+// becomes unknown. Otherwise, the use of this value in comparisons might yield invalid results.
```

Comment:
So marking "a.b.c" unknown makes "a" unknown?  Then is there ever any difference between declaring "a.b.c" to be unknown vs just declaring "a" to be unknown?